bindings/c/test/apitester/TesterTransactionExecutor.cpp

```diff
@@ -0,0 +1,282 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "TesterTransactionExecutor.h"
+#include "TesterUtil.h"
+#include "test/apitester/TesterScheduler.h"
+#include <memory>
+#include <unordered_map>
+#include <mutex>
+#include <atomic>
+
+namespace FdbApiTester {
+
+void ITransactionContext::continueAfterAll(std::shared_ptr<std::vector<Future>> futures, TTaskFct cont) {
+	auto counter = std::make_shared<std::atomic<int>>(futures->size());
+	for (auto& f : *futures) {
+		continueAfter(f, [counter, cont]() {
+			if (--(*counter) == 0) {
+				cont();
+			}
+		});
+	}
+}
+
+class TransactionContext : public ITransactionContext {
+public:
+	TransactionContext(FDBTransaction* tx,
+	                   std::shared_ptr<ITransactionActor> txActor,
+	                   TTaskFct cont,
+	                   const TransactionExecutorOptions& options,
+	                   IScheduler* scheduler)
+	  : options(options), fdbTx(tx), txActor(txActor), contAfterDone(cont), scheduler(scheduler) {}
+
+	Transaction* tx() override { return &fdbTx; }
+	void continueAfter(Future f, TTaskFct cont) override { doContinueAfter(f, cont); }
+	void commit() override {
+		Future f = fdbTx.commit();
+		doContinueAfter(f, [this]() { done(); });
+	}
+	void done() override {
+		TTaskFct cont = contAfterDone;
+		ASSERT(!onErrorFuture);
+		ASSERT(waitMap.empty());
+		delete this;
+		cont();
+	}
+
+private:
+	void doContinueAfter(Future f, TTaskFct cont) {
+		if (options.blockOnFutures) {
+			blockingContinueAfter(f, cont);
+		} else {
+			asyncContinueAfter(f, cont);
+		}
+	}
+
+	void blockingContinueAfter(Future f, TTaskFct cont) {
+		scheduler->schedule([this, f, cont]() mutable {
+			std::unique_lock<std::mutex> lock(mutex);
+			if (!onErrorFuture) {
+				fdb_error_t err = fdb_future_block_until_ready(f.fdbFuture());
+				if (err) {
+					lock.unlock();
+					transactionFailed(err);
+					return;
+				}
+				err = f.getError();
+				if (err) {
+					if (err != error_code_transaction_cancelled) {
+						onErrorFuture = fdbTx.onError(err);
+						fdb_error_t err2 = fdb_future_block_until_ready(onErrorFuture.fdbFuture());
+						if (err2) {
+							lock.unlock();
+							transactionFailed(err2);
+							return;
+						}
+						scheduler->schedule([this]() { handleOnErrorResult(); });
+					}
+				} else {
+					scheduler->schedule([cont]() { cont(); });
+				}
+			}
+		});
+	}
+
+	void asyncContinueAfter(Future f, TTaskFct cont) {
+		std::unique_lock<std::mutex> lock(mutex);
+		if (!onErrorFuture) {
+			waitMap[f.fdbFuture()] = WaitInfo{ f, cont };
+			lock.unlock();
+			fdb_error_t err = fdb_future_set_callback(f.fdbFuture(), futureReadyCallback, this);
```

Comment:
You are right, I made some tests cancelling the transaction before all callbacks complete. I rewrote the implementation of transaction context completely so that it is now using shared references for callbacks. Also made the an explicit state machine for representing different transaction states.